File: openhab/mounts/tmp/openhab/conf/automation/jsr223/python/personal/core_checks.py
from __future__ import unicode_literals
from core.triggers import when
from core.rules import rule
from personal.core_broadcast import broadcast
from core.jsr223.scope import ir, events, things
from core.date import days_between, ZonedDateTime
from personal.core_helpers import get_date
from core import osgi

ITEM_CHANNEL_LINK_REGISTRY = osgi.get_service("org.openhab.core.thing.link.ItemChannelLinkRegistry")
ELAPSED_DAYS = 5

@rule("Core - Check things for offline state.", description="Check things for offline state.", tags=['core', 'core-checks'])
@when("Time cron 0 0/5 * * * ?")
def offline_check(event):
    group = ir.getItem("gCore_Checks_OfflineThings")
    for member in group.allMembers:   
        group.removeMember(member)

    allThings = things.getAll()
    for thing in allThings:
        if (
            thing.getStatusInfo() and 
            thing.getStatusInfo().getStatus().toString() != "ONLINE"
            ) or ( 
            thing.getProperties() and 
            thing.getProperties().get("zwave_lastheal") and
            days_between(get_date(thing.getProperties().get("zwave_lastheal"), "yyyy-MM-dd'T'HH:mm:ssX"), ZonedDateTime.now()) > ELAPSED_DAYS
        ):
            for channel in thing.getChannels():
                for item in ITEM_CHANNEL_LINK_REGISTRY.getLinkedItems(channel.getUID()):
                    if not item.getGroupNames() or str(group.name) not in item.getGroupNames():
                        group.addMember(item)

File: openhab/mounts/tmp/openhab/conf/automation/js/core-checks.js
const { rules, items, things, triggers, osgi, time, actions } = require('openhab')

const ItemChannelLinkRegistry = osgi.getService(
  'org.openhab.core.thing.link.ItemChannelLinkRegistry'
)

const ELAPSED_DAYS = 5
const ZWAVE_DATETIME_FORMAT = time.DateTimeFormatter.ofPattern(
  "yyyy-MM-dd'T'HH:mm:ssX"
)

function scriptLoaded() {
  rules.JSRule({
    name: 'offline_check',
    description: 'Core (JS) - Check things for offline state.',
    tags: ['core', 'core-checks'],
    triggers: [triggers.GenericCronTrigger('0 0/5 * * * ?')],
    execute: (event) => {
      const group = items.getItem('gCore_Checks_OfflineThings')
      for (const member of group.members) {
        member.removeGroups(group)
      }
      const allThings = things.getThings()
      for (const thing of allThings) {
        if (
          thing.status != 'ONLINE' ||
          (thing.rawThing.getProperties() &&
            thing.rawThing.getProperties().get('zwave_lastheal') &&
            time.ZonedDateTime.parse(
              thing.rawThing.getProperties().get('zwave_lastheal'),
              ZWAVE_DATETIME_FORMAT
            ).until(time.ZonedDateTime.now(), time.ChronoUnit.DAYS)) >
            ELAPSED_DAYS
        ) {
          for (const channel of thing.rawThing.getChannels()) {
            for (let item of ItemChannelLinkRegistry.getLinkedItems(
              channel.getUID()
            )) {
              if (!actions.Semantics.isEquipment(item)) {
                item = actions.Semantics.getEquipment(item)
              }

              if (item) {
                items.getItem(item.name).addGroups(group)
              }
            }
          }
        }
      }
    }
  })
}
